File: backend/app/schemas/ai.py
from __future__ import annotations

from uuid import UUID

from pydantic import BaseModel, Field


# ==========================================================
# Request
# ==========================================================

class PredictionRequest(BaseModel):
    """
    Request body for AI prediction.
    """

    upload_id: UUID | None = Field(
        default=None,
        description="ID of the uploaded conversation.",
    )

    conversation: list[str] = Field(
        ...,
        min_length=1,
        description="Conversation messages in chronological order.",
    )


# ==========================================================
# Prediction
# ==========================================================

class PredictionResponse(BaseModel):
    """
    AI prediction response.
    """

    label: str
    probability: float
    confidence: float
    risk_score: float
    message_count: int
    prediction_time_ms: float
    model_version: str


# ==========================================================
# SHAP Explanation
# ==========================================================

class WordExplanationResponse(BaseModel):
    """
    Individual SHAP word/token explanation.
    """

    token: str
    score: float
    normalized_score: float
    importance: str
    color: str


class ExplanationResponse(BaseModel):
    """
    Collection of SHAP word explanations.
    """

    summary: str
    words: list[WordExplanationResponse]


# ==========================================================
# Behavioral Indicators
# ==========================================================

class BehavioralIndicatorResponse(BaseModel):
    """
    Human-readable behavioral indicators detected
    in the conversation.
    """

    indicator_type: str
    title: str
    description: str
    severity: str
    count: int


# ==========================================================
# Prediction Timeline
# ==========================================================

class PredictionTimelinePointResponse(BaseModel):
    """
    AI prediction after each conversation checkpoint.
    """

    message_index: int
    label: str
    probability: float
    confidence: float
    risk_score: float


# ==========================================================
# Explanation Timeline
# ==========================================================

class ExplanationTimelinePointResponse(BaseModel):
    """
    SHAP evidence accumulation across the conversation.
    """

    message_index: int
    risk_score: float


# ==========================================================
# Combined Response
# ==========================================================

class PredictionWithExplanationResponse(BaseModel):
    """
    Complete AI analysis response containing:

    - Prediction
    - SHAP explanation
    - Behavioral indicators
    - Prediction timeline
    - Explanation timeline
    """

    prediction: PredictionResponse

    explanation: ExplanationResponse

    behavioral_indicators: list[
        BehavioralIndicatorResponse
    ]

    prediction_timeline: list[
        PredictionTimelinePointResponse
    ]

    explanation_timeline: list[
        ExplanationTimelinePointResponse
    ]


# ==========================================================
# Model Information
# ==========================================================

class ModelInfoResponse(BaseModel):
    """
    Information about the loaded AI model.
    """

    model_name: str
    model_version: str
    architecture: str
    best_validation_f1: float
    max_sequence_length: int
    device: str
    labels: list[str]

File: frontend/src/types/ai.ts
export interface PredictionRequest {
  upload_id?: string | null;
  conversation: string[];
}

export interface PredictionResponse {
  label: string;
  probability: number;
  confidence: number;
  risk_score: number;
  message_count: number;
  prediction_time_ms: number;
  model_version: string;
}

/* -------------------------------- */
/* SHAP Word Explanation             */
/* -------------------------------- */

export interface WordExplanationResponse {
  token: string;
  score: number;
  normalized_score: number;
  importance: string;
  color: string;
}

export interface ExplanationResponse {
  summary: string;
  words: WordExplanationResponse[];
}

/* -------------------------------- */
/* Behavioral Indicators             */
/* -------------------------------- */

export interface BehavioralIndicatorResponse {
  indicator_type: string;
  title: string;
  description: string;
  severity: string;
  count: number;
}

/* -------------------------------- */
/* Early Prediction Timeline         */
/* -------------------------------- */

export interface PredictionTimelinePoint {
  message_index: number;
  label: string;
  probability: number;
  confidence: number;
  risk_score: number;
}

/* -------------------------------- */
/* SHAP Explanation Timeline         */
/* -------------------------------- */

export interface ExplanationTimelinePoint {
  message_index: number;
  risk_score: number;
}

/* -------------------------------- */
/* /predict/explain Response         */
/* -------------------------------- */

export interface PredictionWithExplanationResponse {
  prediction: PredictionResponse;

  explanation: ExplanationResponse;

  behavioral_indicators: BehavioralIndicatorResponse[];

  prediction_timeline: PredictionTimelinePoint[];

  explanation_timeline: ExplanationTimelinePoint[];
}

/* -------------------------------- */
/* Model Information                 */
/* -------------------------------- */

export interface ModelInfoResponse {
  model_name: string;
  model_version: string;
  architecture: string;
  best_validation_f1: number;
  max_sequence_length: number;
  device: string;
  labels: string[];
}